onnxruntime/core/providers/coreml/model/model.mm

```diff
@@ -0,0 +1,319 @@
+  const std::unordered_map<std::string, onnxruntime::coreml::OnnxTensorData>* inputs_;
+  NSSet* featureNames_;
+}
+
+- (instancetype)initWithInputs:(const std::unordered_map<std::string, onnxruntime::coreml::OnnxTensorData>&)inputs;
+- (MLFeatureValue*)featureValueForName:(NSString*)featureName API_AVAILABLE_OS_VERSIONS;
+- (NSSet<NSString*>*)featureNames;
+
+@end
+
+// Execution for a CoreML model, it performs
+// 1. Compile the model by given path for execution
+// 2. Predict using given OnnxTensorFeatureProvider input and copy the output data back ORT
+// 3. The compiled model will be removed in dealloc or removed using cleanup function
+@interface CoreMLExecution : NSObject {
+  NSString* coreml_model_path_;
+  NSString* compiled_model_path_;
+}
+
+- (instancetype)initWithPath:(const std::string&)path;
+- (void)cleanup;
+- (void)dealloc;
+- (onnxruntime::common::Status)loadModel API_AVAILABLE_OS_VERSIONS;
+- (onnxruntime::common::Status)
+    predict:(const std::unordered_map<std::string, onnxruntime::coreml::OnnxTensorData>&)inputs
+    outputs:(const std::unordered_map<std::string, onnxruntime::coreml::OnnxTensorData>&)outputs
+    API_AVAILABLE_OS_VERSIONS;
+
+@property MLModel* model API_AVAILABLE_OS_VERSIONS;
+
+@end
+
+@implementation OnnxTensorFeatureProvider
+
+- (instancetype)initWithInputs:(const std::unordered_map<std::string, onnxruntime::coreml::OnnxTensorData>&)inputs {
+  self = [super init];
+  inputs_ = &inputs;
+  return self;
+}
+
+- (nonnull NSSet<NSString*>*)featureNames {
+  if (featureNames_ == nil) {
+    NSMutableArray* names = [[NSMutableArray alloc] init];
+    for (const auto& input : *inputs_) {
+      [names addObject:[NSString stringWithCString:input.first.c_str()
+                                          encoding:[NSString defaultCStringEncoding]]];
+    }
+
+    featureNames_ = [NSSet setWithArray:names];
+  }
+
+  return featureNames_;
+}
+
+- (nullable MLFeatureValue*)featureValueForName:(nonnull NSString*)featureName {
+  auto it = inputs_->find([featureName cStringUsingEncoding:NSUTF8StringEncoding]);
+  if (it != inputs_->end()) {
+    auto& input = it->second;
+    NSMutableArray* shape = [[NSMutableArray alloc] init];
+    for (const auto dim : input.tensor_info.shape) {
+      [shape addObject:[NSNumber numberWithLongLong:dim]];
+    }
+
+    NSMutableArray* strides = [[NSMutableArray alloc] init];
+    int64_t stride = 1;
+    for (int i = static_cast<int>(input.tensor_info.shape.size()) - 1; i >= 0; i--) {
+      [strides insertObject:[NSNumber numberWithLongLong:stride]
+                    atIndex:0];
+
+      stride *= input.tensor_info.shape[i];
+    }
+
+    MLMultiArrayDataType data_type = MLMultiArrayDataTypeFloat32;
+    if (input.tensor_info.data_type != ONNX_NAMESPACE::TensorProto_DataType_FLOAT) {
+      LOGS_DEFAULT(ERROR) << "Input data type is not float, actual type: "
+                          << input.tensor_info.data_type;
```

Comment:
IExecutionProvider::GetLogger can be used to get the session logger which is preferable to using the default logger.

Reply:
Is there a good way to push the logger into ModelBuilder and other helper functions?


---
In reply to: [563492954](https://github.com/microsoft/onnxruntime/pull/6392#discussion_r563492954) [](ancestors = 563492954)